content/docs/reference/design-docs/editoast-errors/_index.en.md

```diff
@@ -0,0 +1,320 @@
+- Each endpoint must provide all its error cases in the OpenAPI. (How the frontend will consume them is another problem that we'll have to deal with.)
+- The `context` field of the error is populated in the views.
+- Errors can be handled in a generic matter (for situations where it makes some sense to do so).
+  - I.e.: some form of downcasting is available.
+
+# New system
+
+- Rely on `thiserror` everywhere.
+- Keep the `trait EditoastError` but only implement it for errors defined in `views`.
+- Rework `derive(EditoastError)` to interface it with `derive(thiserror::Error)`.
+- The `context` is empty by default but can be provided by the `impl EditoastError`. The macro is also able to take context providers.
+- `EditoastError`'s `#[source]`, `#[from]`, `source` and `backtrace` fields are never serialized, unless explicitely provided. This shouldn't be the case as it exposes editoast internals at the API level.
+- All errors must `impl Debug`.
+- `derive(EditoastError)` generates the `impl utoipa::ToSchema for T`.
+- Errors **do not** require to `impl Serialize`.
+- `impl Serialize for T where T: EditoastError` however.
+- Error cases that will be used repeatedly are defined as a `struct` but still `derive(thiserror::Error)`.
+- The `error_type` of each variant is generated by the macro at the format `ErrorTypeName::VariantName`, but can be provided explicitely if conflicts arise.
+
+## Nominal case
+
+```rust
+// in mod views;
+
+fn get_resource(key: Key) -> Result<Resource, GetError> {
+    todo!()
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum GetError {
+    #[error("ID not found")]
+    IdNotFound { id: u64 },
+    #[error("name not found")]
+    NameNotFound { name: String }
+}
+
+fn process_resource(resource: Resource) -> Result<Computation, ProcessingError> {
+    todo!()
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ProcessingError {
+    #[error("Resource is invalid")]
+    InvalidResource { resource: Resource },
+    #[error("Resource is too old")]
+    OutdatedResource { resource: Resource }
+}
+
+#[utoipa::path(
+    ...,
+    responses(
+        (status = 200, body = Computation),
+        (status = 400, body = EndpointError)
+    )
+)]
+async fn endpoint(Path(key): Path<Key>) -> Result<Json<Computation>, EndpointError> {
+    let resource = get_resource(key)?;
+    let value = process_resource(resource)?;
+    Ok(Json(value))
+}
+
+#[derive(Debug, thiserror::Error, EditoastError)]
+pub enum EndpointError {
+    #[error("Resource not found")]
+    #[editoast_error(user)] // <=> status = 400
+    ResourceNotFound(#[from] GetError),
+
+    #[editoast_error(internal)] // <=> status = 500, default, overrides the message to "Internal error - please contact your administrator"
+    ProcessingFailed(#[from] ProcessingError)
+}
+```
+
+## Share errors between crates
+
+Since we require no other constraint that `impl Error` for composition, it's easy to nest errors using `thiserror`.
```

Comment:
even bigger nitpick, but `core::error::Error` is stable now.

```suggestion
Since we require no other constraint that `impl core::error::Error` for composition, it's easy to nest errors using `thiserror`.
```